presto-hive/src/main/java/com/facebook/presto/hive/metastore/glue/GlueHiveMetastore.java

```diff
@@ -0,0 +1,703 @@
+    {
+        List<String> partitionNames = new ArrayList<>();
+        List<String> partitionKeys = partitionColumns.stream().map(Column::getName).collect(toList());
+        partitions.stream().forEach(p -> partitionNames.add(toPartitionName(partitionKeys, p.getValues())));
+        return partitionNames;
+    }
+
+    /**
+     * @param partitionNames is a list of full partition names
+     * ex: partition keys = ['a', 'b']
+     * partition names:
+     * ['a=1/b=2', 'a=2/b=2']
+     * @return is a mapping of partition name to partition object.
+     */
+    @Override
+    public Map<String, Optional<Partition>> getPartitionsByNames(String databaseName, String tableName, List<String> partitionNames)
+    {
+        requireNonNull(partitionNames, "partitionNames is null");
+        if (partitionNames.isEmpty()) {
+            return ImmutableMap.of();
+        }
+
+        List<Partition> partitions = batchGetPartition(databaseName, tableName, partitionNames);
+
+        Map<String, List<String>> partitionNameToPartitionValuesMap = partitionNames.stream()
+                .collect(toMap(identity(), HiveUtil::toPartitionValues));
+        Map<List<String>, Partition> partitionValuesToPartitionMap = partitions.stream()
+                .collect(toMap(Partition::getValues, identity()));
+
+        ImmutableMap.Builder<String, Optional<Partition>> resultBuilder = ImmutableMap.builder();
+        for (Map.Entry<String, List<String>> entry : partitionNameToPartitionValuesMap.entrySet()) {
+            Partition partition = partitionValuesToPartitionMap.get(entry.getValue());
+            resultBuilder.put(entry.getKey(), Optional.ofNullable(partition));
+        }
+        return resultBuilder.build();
+    }
+
+    private List<Partition> batchGetPartition(String databaseName, String tableName, List<String> partitionNames)
+    {
+        try {
+            List<PartitionValueList> partitionValueLists = partitionNames.stream()
+                    .map(p -> new PartitionValueList().withValues(toPartitionValues(p))).collect(toList());
+
+            List<List<PartitionValueList>> batchedPartitionValueLists = Lists.partition(partitionValueLists, BATCH_GET_PARTITION_MAX_PAGE_SIZE);
+            List<Future<BatchGetPartitionResult>> batchGetPartitionFutures = new ArrayList<>();
+            List<Partition> result = new ArrayList<>();
+
+            for (List<PartitionValueList> partitions : batchedPartitionValueLists) {
+                batchGetPartitionFutures.add(glueClient.batchGetPartitionAsync(new BatchGetPartitionRequest()
+                        .withDatabaseName(databaseName)
+                        .withTableName(tableName)
+                        .withPartitionsToGet(partitions)));
+            }
+
+            for (Future<BatchGetPartitionResult> future : batchGetPartitionFutures) {
+                BatchGetPartitionResult batchResult = future.get();
```

Comment:
We should probably have a timeout on the `Future.get()` call, in case the client has a bug where the future never completes.

Reply:
The client already have a [default connection & socket timeout](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-core/src/main/java/com/amazonaws/ClientConfiguration.java#L42), would that also work or should the ```Future.get()``` timeout  be set as well?

Reply:
We've seen many issues where a future does not complete for whatever reason, then the thread and query will hang forever (or until the query is cancelled by the user which will interrupt the thread). So while it might not be strictly necessary, it's safer to pass a timeout to the `get()` (which we hope will never be hit).